rust/sdk/src/client.rs

```diff
@@ -388,6 +316,356 @@ pub(crate) fn to_lines(stream: impl Stream<Item = Result<Bytes, reqwest::Error>>
+    /// Add events for publishing, from an iterable.
+    ///
+    /// Subsequent calls will add the events rather than replacing them.
+    pub fn events<E: IntoIterator<Item = impl Into<PublishEvent>>>(mut self, events: E) -> Self {
+        if let Self::Initial { ref mut request, .. } = self {
+            request.data.extend(events.into_iter().map(Into::into));
+        }
+        self
+    }
+}
+
+impl<'a> Future for Publish<'a> {
+    type Output = anyhow::Result<PublishResponse>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> std::task::Poll<Self::Output> {
+        let this = self.get_mut();
+        loop {
+            *this = match replace(this, Publish::Empty) {
+                Publish::Initial { client, request } => {
+                    let publish_response = async move {
+                        let publish_url = client.events_url("publish");
+                        let response = client.do_request(|c| c.post(publish_url).json(&request)).await?;
+                        let body = response.bytes().await?;
+                        Ok(serde_json::from_slice::<PublishResponse>(&body)?)
+                    };
+                    Publish::Pending(publish_response.boxed())
+                }
+                Publish::Pending(mut publish_response_future) => {
+                    let polled = publish_response_future.poll_unpin(cx);
+                    if polled.is_pending() {
+                        *this = Publish::Pending(publish_response_future);
+                    }
+                    return polled;
+                }
+                Publish::Empty => panic!("Polling a terminated Publish future"),
+            };
+        }
+    }
+}
+
+impl<'a> FusedFuture for Publish<'a> {
+    fn is_terminated(&self) -> bool {
+        if let Publish::Empty = self {
+            true
+        } else {
+            false
+        }
+    }
+}
+
+pub enum Query<'a> {
+    Initial {
+        client: &'a ActyxClient,
+        request: QueryRequest,
+    },
+    Pending(BoxFuture<'a, anyhow::Result<BoxStream<'a, QueryResponse>>>),
+    Empty,
+}
+
+impl<'a> Query<'a> {
+    fn new<Q: Into<String>>(client: &'a ActyxClient, query: Q) -> Self {
+        Self::Initial {
+            client,
+            request: QueryRequest {
+                query: query.into(),
+                lower_bound: Some(OffsetMap::empty()),
+                upper_bound: None,
+                order: Order::Asc,
+            },
+        }
+    }
+
+    /// Add a lower bound to the query.
+    ///
+    /// The lower bound limits the start of the query events.
+    /// As an example, consider the following (example) events:
+    /// ```json
+    /// { "lamport": 1, "event": { "temperature": 10 } }
+    /// { "lamport": 3, "event": { "temperature": 12 } }
+    /// { "lamport": 14, "event": { "temperature": 9 } }
+    /// ```
+    /// If you set the lower bound to `10`, only the last event will be returned.
```

Comment:
This explanation is incorrect: the lower bound uses Offsets, not timestamps, so the interpretation isn’t that easy. It is also impossible to just make up an OffsetMap because you’d need some StreamIds that you cannot guess. An example for this method would need to involve a call to `client.offsets()` to get an OffsetMap, I think.

Reply:
Should we receive offsets from Actyx API exclusively?